Main.java

```diff
@@ -0,0 +1,153 @@
+
+		grid[0][0] = 'A';
+		grid[0][1] = 'T';
+		grid[1][0] = 'A';
+		grid[1][1] = 'C';
+		grid[2][0] = 'R';
+		grid[2][1] = 'D';
+
+		dict = new Dictionary(new String[] {});
+
+		System.out.print("\tResult: ");
+		words = findWords(grid, dict);
+		for (String w : words) {
+			System.out.print(w + " ");
+		}
+
+		System.out.println("\n4. Empty grid and empty dictionary. Shouldn't find any word");
+
+		grid = new Character[100][100];
+		dict = new Dictionary(new String[] {});
+
+		System.out.print("\tResult: ");
+		words = findWords(grid, dict);
+		for (String w : words) {
+			System.out.print(w + " ");
+		}
+
+		System.out.println("\n5. Grid without words from dictionary. Shouldn't find any word");
+
+		grid = new Character[100][100];
+		for (int i = 0; i < 100; i++) {
+			for (int j = 0; j < 100; j++) {
+				grid[i][j] = (i + j % 2 == 0) ? 'A' : 'B';
+			}
+		}
+
+		dict = new Dictionary(new String[] { "AAAAAAAAAAAAAAAAAAAAAAAAAAAC", "ABABABABABABAAAAAAAAAAAAAAAABBBBAAAAX" });
+
+		System.out.print("\tResult: ");
+		words = findWords(grid, dict);
+		for (String w : words) {
+			System.out.print(w + " ");
+		}
+
+		System.out.println(
+				"\n5. Big grid with words from dictionary. Should find AAAAAAAAAAAAAAAAAAAAAA BBBBBBBBBBBBBBBBBBBBBBBBBBB ABABABABABABABABAB");
+
+		grid = new Character[100][100];
+		for (int i = 0; i < 100; i++) {
+			for (int j = 0; j < 100; j++) {
+				grid[i][j] = ((i + j) % 2 == 0) ? 'A' : 'B';
+			}
+		}
+
+		dict = new Dictionary(
+				new String[] { "AAAAAAAAAAAAAAAAAAAAAA", "BBBBBBBBBBBBBBBBBBBBBBBBBBB", "ABABABABABABABABAB" });
+
+		System.out.print("\tResult: ");
+		words = findWords(grid, dict);
+		for (String w : words) {
+			System.out.print(w + " ");
+		}
+	}
+
+	/**
+	 * This method finds words from a given dictionary 
+	 * contained in a two dimensional grid
+	 * 
+	 * @param grid - Two dimensional array containing chars
+	 * @param dict - Dictionary with valid words
+	 * @return - Set of the valid words from the given grid
+	 */
+	public static Set<String> findWords(Character [][] grid, Dictionary dict) {
+		if (grid == null || dict == null) {
+			throw new IllegalArgumentException();
+		}
+
+		Set<String> result = new HashSet<String>();
+		int height = grid.length;
+		int width = grid[0].length;
+
+		for (int i = 0; i < width; i++) {
+			for (int j = 0; j < height; j++) {
+				Point currentPoint = new Point(i, j, width, height);
+				result.addAll(recursiveFindWords("", currentPoint, grid, dict));
+			}
+		}
+		return result;
+	}
+
+	public static Set<String> recursiveFindWords(String currentWord, Point point, Character [][] grid, Dictionary dict) {
+		Set<String> result = new HashSet<String>();
+
+		if (!dict.isPrefix(currentWord)) {
+			return result;
+		} else {
+			char letter = grid[point.y][point.x];
+			grid[point.y][point.x] = null;
+			currentWord = currentWord + letter;
+			
+			if (dict.isWord(currentWord)) {
+				result.add(currentWord);
+				dict.removeWord(currentWord);
```

Comment:
@Thomas-de-Rivaz Instead of adding the words again is it a good idea to make a copy of the dictionary and modify the local copy?